File: scripts/create_work_orders_bucket.py
#!/usr/bin/env python3
"""
Script to create the work_orders bucket in InfluxDB if it doesn't exist.
"""

import os
import requests
from influxdb_client import InfluxDBClient, BucketsApi, BucketRetentionRules

# InfluxDB configuration
INFLUXDB_URL = os.getenv("INFLUXDB_URL", "http://localhost:8086")
INFLUXDB_TOKEN = os.getenv("INFLUXDB_TOKEN", "my-super-secret-auth-token")
INFLUXDB_ORG = os.getenv("INFLUXDB_ORG", "myorg")
BUCKET_NAME = "work_orders"

def create_bucket_if_not_exists():
    """Create the work_orders bucket if it doesn't exist."""
    try:
        client = InfluxDBClient(
            url=INFLUXDB_URL,
            token=INFLUXDB_TOKEN,
            org=INFLUXDB_ORG
        )
        
        buckets_api = BucketsApi(client)
        
        # Check if bucket exists
        buckets = buckets_api.find_buckets()
        existing_bucket = None
        
        for bucket in buckets.buckets:
            if bucket.name == BUCKET_NAME:
                existing_bucket = bucket
                break
        
        if existing_bucket:
            print(f"✅ Bucket '{BUCKET_NAME}' already exists")
            print(f"   ID: {existing_bucket.id}")
            print(f"   Retention: {existing_bucket.retention_rules}")
        else:
            print(f"📦 Creating bucket '{BUCKET_NAME}'...")
            
            # Create bucket with no retention (infinite retention)
            retention_rules = BucketRetentionRules(type="expire", every_seconds=0)
            bucket = buckets_api.create_bucket(
                bucket_name=BUCKET_NAME,
                org=INFLUXDB_ORG,
                retention_rules=retention_rules
            )
            
            print(f"✅ Bucket '{BUCKET_NAME}' created successfully")
            print(f"   ID: {bucket.id}")
        
        client.close()
        return True
        
    except Exception as e:
        print(f"❌ Error creating bucket: {e}")
        return False

if __name__ == "__main__":
    print(f"🔍 Checking InfluxDB bucket: {BUCKET_NAME}")
    print(f"   URL: {INFLUXDB_URL}")
    print(f"   Org: {INFLUXDB_ORG}\n")
    
    success = create_bucket_if_not_exists()
    exit(0 if success else 1)

File: scripts/create-work-orders-bucket.js
const https = require('https');

const INFLUXDB_URL = process.env.INFLUXDB_URL || 'https://influxtest.wisermachines.com';
const INFLUXDB_TOKEN = process.env.INFLUXDB_TOKEN || '1MrRJ8q-zSnlt9HRZMeY5YNhOQZWbi6Xk-oU6pFFTSbJRv4V32cTJutWMJota0r6t_F6N5zXOfE6IXHYmcUk4Q==';
const INFLUXDB_ORG = process.env.INFLUXDB_ORG || 'wisermachines';
const BUCKET_NAME = 'work_orders';

// First, get the org ID
const getOrgId = () => {
  return new Promise((resolve, reject) => {
    const url = `${INFLUXDB_URL}/api/v2/orgs?org=${INFLUXDB_ORG}`;
    const urlObj = new URL(url);
    
    const options = {
      hostname: urlObj.hostname,
      port: urlObj.port || 443,
      path: urlObj.pathname + urlObj.search,
      method: 'GET',
      headers: {
        'Authorization': `Token ${INFLUXDB_TOKEN}`,
        'Content-Type': 'application/json',
      },
    };

    const req = https.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => {
        if (res.statusCode === 200) {
          const orgs = JSON.parse(data).orgs || [];
          if (orgs.length > 0) {
            resolve(orgs[0].id);
          } else {
            reject(new Error('Organization not found'));
          }
        } else {
          reject(new Error(`Failed to get org ID: ${res.statusCode} ${data}`));
        }
      });
    });

    req.on('error', reject);
    req.end();
  });
};

// Create the bucket
const createBucket = (orgId) => {
  return new Promise((resolve, reject) => {
    const url = `${INFLUXDB_URL}/api/v2/buckets`;
    const urlObj = new URL(url);
    
    const bucketData = {
      name: BUCKET_NAME,
      orgID: orgId,
      retentionRules: [], // No retention (infinite)
    };

    const options = {
      hostname: urlObj.hostname,
      port: urlObj.port || 443,
      path: urlObj.pathname,
      method: 'POST',
      headers: {
        'Authorization': `Token ${INFLUXDB_TOKEN}`,
        'Content-Type': 'application/json',
      },
    };

    const req = https.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => {
        if (res.statusCode === 201 || res.statusCode === 200) {
          console.log(`✅ Bucket "${BUCKET_NAME}" created successfully!`);
          console.log(JSON.parse(data));
          resolve(JSON.parse(data));
        } else {
          if (data.includes('already exists')) {
            console.log(`ℹ️  Bucket "${BUCKET_NAME}" already exists`);
            resolve({ exists: true });
          } else {
            reject(new Error(`Failed to create bucket: ${res.statusCode} ${data}`));
          }
        }
      });
    });

    req.on('error', reject);
    req.write(JSON.stringify(bucketData));
    req.end();
  });
};

// Main execution
(async () => {
  try {
    console.log(`Creating bucket "${BUCKET_NAME}" in org "${INFLUXDB_ORG}"...`);
    const orgId = await getOrgId();
    console.log(`Found org ID: ${orgId}`);
    await createBucket(orgId);
  } catch (error) {
    console.error('❌ Error:', error.message);
    process.exit(1);
  }
})();

